From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sat, 9 Aug 2025 00:28:53 -0500
Subject: [PATCH] Fish Schedule Async Task But Wait Outside Lock

This patch is purely experimental, you can call it a proof of concept.
Given the nature of PWT in Fish, where async tasks can still be async as long as the world is not ticking, there's a
  scenario where, if a chunk is not loaded and cannot be loaded async, it will schedule its load at the end of the world
  tick (moonrise behavior) and wait. However, since the async task acquired the lock, then the world cannot acquire it
  to execute the sync chunk load at the end of its tick. Yay, deadlock!
The original solution was to schedule the task to always be enqueued on the main thread, however this means that a
  task that 99% of times will perform successfully async, will be turned sync for the 1% scenarios where a deadlock happens.
This patch (as proof of concept) is meant to still perform the task async if the thread managed to acquire the lock, but
  perform the wait outside the lock, which allows the main thread to process it at the end of its tick.
Only downside is that the affected code its spreaded in three (3) files, which means I now have to maintain those...

diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index fcc965f4d5e8ef2e6879033eae891b8a41288481..0c346549412cc26fdbee9dbab6e9752a54748125 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -101,7 +101,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         return this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
     }
 
-    private ChunkAccess syncLoad(final int chunkX, final int chunkZ, final ChunkStatus toStatus) {
+    private ChunkAccess syncLoad(final int chunkX, final int chunkZ, final ChunkStatus toStatus) { // Fish - diff on change
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
         final CompletableFuture<ChunkAccess> completable = new CompletableFuture<>();
         chunkTaskScheduler.scheduleChunkLoad(
@@ -130,7 +130,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         return ret;
     }
 
-    private ChunkAccess getChunkFallback(final int chunkX, final int chunkZ, final ChunkStatus toStatus,
+    private ChunkAccess getChunkFallback(final int chunkX, final int chunkZ, final ChunkStatus toStatus, // Fish - diff on change
                                          final boolean load) {
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
         final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager chunkHolderManager = chunkTaskScheduler.chunkHolderManager;
@@ -317,7 +317,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @Nullable
     @Override
-    public ChunkAccess getChunk(int x, int z, ChunkStatus chunkStatus, boolean requireChunk) {
+    public ChunkAccess getChunk(int x, int z, ChunkStatus chunkStatus, boolean requireChunk) { // Fish - diff on change
         // Paper start - rewrite chunk system
         if (chunkStatus == ChunkStatus.FULL) {
             final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
@@ -563,6 +563,69 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         // Pufferfish end - Optimize mob spawning
     }
 
+    private CompletableFuture<ChunkAccess> fish$syncLoad(final int chunkX, final int chunkZ, final ChunkStatus toStatus) {
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
+        final CompletableFuture<ChunkAccess> completable = new CompletableFuture<>();
+        chunkTaskScheduler.scheduleChunkLoad(
+            chunkX, chunkZ, toStatus, true, ca.spottedleaf.concurrentutil.util.Priority.BLOCKING,
+            completable::complete
+        );
+
+        if (!completable.isDone() && chunkTaskScheduler.hasShutdown()) {
+            throw new IllegalStateException(
+                "Chunk system has shut down, cannot process chunk requests in world '" + ca.spottedleaf.moonrise.common.util.WorldUtil.getWorldName(this.level) + "' at "
+                    + "(" + chunkX + "," + chunkZ + ") status: " + toStatus
+            );
+        }
+
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.level, chunkX, chunkZ)) {
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.pushChunkWait(this.level, chunkX, chunkZ);
+            this.mainThreadProcessor.managedBlock(completable::isDone);
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler.popChunkWait();
+        }
+
+        return completable;
+    }
+    private CompletableFuture<ChunkAccess> fish$getChunkFallback(final int chunkX, final int chunkZ, final ChunkStatus toStatus,
+                                                                 final boolean load) {
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler chunkTaskScheduler = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)this.level).moonrise$getChunkTaskScheduler();
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkHolderManager chunkHolderManager = chunkTaskScheduler.chunkHolderManager;
+
+        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder currentChunk = chunkHolderManager.getChunkHolder(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(chunkX, chunkZ));
+
+        final ChunkAccess ifPresent = currentChunk == null ? null : currentChunk.getChunkIfPresent(toStatus);
+
+        if (ifPresent != null && (toStatus != ChunkStatus.FULL || currentChunk.isFullChunkReady())) {
+            return CompletableFuture.completedFuture(ifPresent);
+        }
+
+        final ca.spottedleaf.moonrise.common.PlatformHooks platformHooks = ca.spottedleaf.moonrise.common.PlatformHooks.get();
+
+        if (platformHooks.hasCurrentlyLoadingChunk() && currentChunk != null) {
+            final ChunkAccess loading = platformHooks.getCurrentlyLoadingChunk(currentChunk.vanillaChunkHolder);
+            if (loading != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) {
+                return CompletableFuture.completedFuture(loading);
+            }
+        }
+
+        return load ? this.fish$syncLoad(chunkX, chunkZ, toStatus) : null;
+    }
+    public CompletableFuture<ChunkAccess> fish$getChunk(int x, int z, ChunkStatus chunkStatus, boolean requireChunk) {
+        // Paper start - rewrite chunk system
+        if (chunkStatus == ChunkStatus.FULL) {
+            final LevelChunk ret = this.fullChunks.get(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(x, z));
+
+            if (ret != null) {
+                return CompletableFuture.completedFuture(ret);
+            }
+
+            return requireChunk ? this.fish$getChunkFallback(x, z, chunkStatus, requireChunk) : null;
+        }
+
+        return this.fish$getChunkFallback(x, z, chunkStatus, requireChunk);
+        // Paper end - rewrite chunk system
+    }
+
     private void broadcastChangedChunks(ProfilerFiller profiler) {
         profiler.push("broadcast");
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 16d9a2573fe7b7022f139824baeac4b3ffb180dc..ed42a44aef5b1ad54992ed0aa6c4e74f141c0025 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -992,12 +992,12 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
         return y < -20000000 || y >= 20000000;
     }
 
-    public final LevelChunk getChunkAt(BlockPos pos) { // Paper - help inline
+    public final LevelChunk getChunkAt(BlockPos pos) { // Paper - help inline // Fish - diff on change
         return this.getChunk(SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getZ()));
     }
 
     @Override
-    public final LevelChunk getChunk(int chunkX, int chunkZ) { // Paper - final to help inline
+    public final LevelChunk getChunk(int chunkX, int chunkZ) { // Paper - final to help inline // Fish - diff on change
         // Paper start - Perf: make sure loaded chunks get the inlined variant of this function
         net.minecraft.server.level.ServerChunkCache cps = ((ServerLevel)this).getChunkSource();
         LevelChunk ifLoaded = cps.getChunkAtIfLoadedImmediately(chunkX, chunkZ);
@@ -1007,6 +1007,19 @@ public abstract class Level implements LevelAccessor, UUIDLookup<Entity>, AutoCl
         return (LevelChunk) cps.getChunk(chunkX, chunkZ, ChunkStatus.FULL, true); // Paper - avoid a method jump
         // Paper end - Perf: make sure loaded chunks get the inlined variant of this function
     }
+    public final java.util.concurrent.CompletableFuture<ChunkAccess> fish$getChunk(int chunkX, int chunkZ) { // Paper - final to help inline
+        // Paper start - Perf: make sure loaded chunks get the inlined variant of this function
+        net.minecraft.server.level.ServerChunkCache cps = ((ServerLevel)this).getChunkSource();
+        LevelChunk ifLoaded = cps.getChunkAtIfLoadedImmediately(chunkX, chunkZ);
+        if (ifLoaded != null) {
+            return java.util.concurrent.CompletableFuture.completedFuture(ifLoaded);
+        }
+        return cps.fish$getChunk(chunkX, chunkZ, ChunkStatus.FULL, true); // Paper - avoid a method jump
+        // Paper end - Perf: make sure loaded chunks get the inlined variant of this function
+    }
+    public final java.util.concurrent.CompletableFuture<ChunkAccess> fish$getChunkAt(BlockPos pos) { // Paper - help inline
+        return this.fish$getChunk(SectionPos.blockToSectionCoord(pos.getX()), SectionPos.blockToSectionCoord(pos.getZ()));
+    }
 
     // Paper start - if loaded
     @Nullable
